src/utils/AssetCache.ts

```diff
@@ -0,0 +1,147 @@
+import * as FileSystem from 'expo-file-system'
+import AsyncStorage from '@react-native-community/async-storage'
+
+// Base directory where all cached files are kept
+const CACHE_KEY = 'FLARE-ASSETS-CACHE'
+const BASE_DIR = FileSystem.documentDirectory + `asset-caches/`
+
+export default class AssetCache {
+  // Internal reference of remote URL to local file
+  static fileMap = {}
+
+  /**
+   * Setup the the system requirements needed for AssetCache to work.
+   * Call only once in your applications runtime.
+   */
+  static async construct() {
+    // Create directory if doesn't exist
+    FileSystem.readDirectoryAsync(BASE_DIR).catch(() =>
+      FileSystem.makeDirectoryAsync(BASE_DIR),
+    )
+
+    // Instantaite the URL => URI map
+    const rawFileMap = await AsyncStorage.getItem(CACHE_KEY)
+    AssetCache.fileMap = rawFileMap != null ? JSON.parse(rawFileMap) : {}
+  }
+
+  /**
+   * Clean up any download assets after an experiment is finished
+   */
+  static async deconstruct() {
+    await FileSystem.deleteAsync(BASE_DIR)
+    await AsyncStorage.removeItem(CACHE_KEY)
+  }
+
+  /**
+   * Update the internal cache maping remote to local files
+   */
+  private static async syncCache() {
+    // Convert JSON to string
+    const rawFileMap = JSON.stringify(AssetCache.fileMap)
+
+    // Save string into storage
+    return await AsyncStorage.setItem(CACHE_KEY, rawFileMap)
+  }
+
+  /**
+   * Generate an appropriate path for a remote file
+   *
+   * @param url The url of the remote file
+   */
+  private static generateFileName(url: string): string {
+    // Utility method used to generate unqique strings
+    // https://lowrey.me/implementing-javas-string-hashcode-in-javascript/
+    const hashCode = function (str: string) {
+      let hash = 0
+      for (let i = 0; i < str.length; i++) {
+        hash += Math.pow(str.charCodeAt(i) * 31, str.length - i)
+        hash = hash & hash // Convert to 32bit integer
+      }
+      return hash
+    }
+
+    return BASE_DIR + hashCode(url)
+  }
+
+  /**
+   * This method allows you to download a remote file for use later.
+   *
+   * @param url URL for the remote file you want to cache
+   */
+  static async cacheFile(url: string): Promise<string | null> {
+    try {
+      // Download file async
+      const { uri } = await FileSystem.downloadAsync(
+        url,
+        AssetCache.generateFileName(url),
+        {},
+      )
+
+      // Update internal map for saved files
+      AssetCache.fileMap[url] = uri
+
+      // Update mapping cache
+      AssetCache.syncCache()
+
+      // Return the new local file
+      return uri
+    } catch (e) {
+      console.error(e)
+      return null
+    }
```

Comment:
If caching fails for any reason, is it safe for the app to continue execution? There should probably be a way to handle that situation.